src/solver/conv_hip_implicit_gemm_fwd_xdlops.cpp

```diff
@@ -0,0 +1,256 @@
+    return false;
+#else
+    if(miopen::IsDisabled(MIOPEN_DEBUG_CONV_IMPLICIT_GEMM_HIP_FWD_XDLOPS{}))
+        return false;
+    if(miopen::IsEnabled(MIOPEN_DEBUG_CONVOLUTION_DETERMINISTIC{}))
+        return false;
+    if(!ctx.use_hip_kernels)
+        return false;
+    if(!ctx.IsInt8()) // check input and weights
+        return false;
+    if(!(ctx.conv_problem.GetOutDataType() == miopenInt8))
+        return false;
+    if(!ctx.direction.IsForward())
+        return false;
+    if(!ctx.Is2d())
+        return false;
+    if(ctx.GetStream().GetDeviceName() != "gfx908")
+        return false;
+    if(!ctx.IsLayoutNHWC())
+        return false;
+
+    std::vector<DeviceConvFwdPtr_t> conv_ptrs;
+    add_device_conv2d_fwd_xdl_nhwc_kyxc_nhwk_int8_instances_t(conv_ptrs);
+    assert(!conv_ptrs.empty());
+    const auto args = CKArgs{ctx};
+    if(!std::all_of(args.strides.begin(), args.strides.end(), [&](auto x) { return x == 1; }))
+        return false;
+
+    for(auto& conv_ptr : conv_ptrs)
+    {
+        auto argument_ptr = conv_ptr.MakeArgumentPointer(nullptr,
+                                                         nullptr,
+                                                         nullptr,
+                                                         args.N,
+                                                         args.K,
+                                                         args.C,
+                                                         args.input,
+                                                         args.filter,
+                                                         args.input,
+                                                         args.strides,
+                                                         args.dilation,
+                                                         args.lPadding,
+                                                         args.rPadding);
+        if(conv_ptr.IsSupportedArgument(argument_ptr.get()))
+            return true;
+    }
+    return false;
+#endif
+}
+
+ConvSolution ConvHipImplicitGemmFwdXdlops::GetSolution(
+    const ConvolutionContext& ctx, const PerformanceConfigHipImplicitGemmFwdXdlops& config) const
+{
+#if !MIOPEN_BACKEND_HIP
+    std::ignore = ctx;
+    std::ignore = config;
+    return {};
+#else
+    ConvSolution result;
+    const auto args        = CKArgs{ctx};
+    result.invoker_factory = [=](const std::vector<Kernel>& kernels) {
+        std::ignore = kernels;
+        return [=](const Handle& handle, const AnyInvokeParams& primitive_parameters) {
+            std::vector<DeviceConvFwdPtr_t> conv_ptrs;
+            add_device_conv2d_fwd_xdl_nhwc_kyxc_nhwk_int8_instances_t(conv_ptrs);
+            auto& conv_ptr       = conv_ptrs.at(config.index);
+            const auto& data_ctx = primitive_parameters.CastTo<conv::DataInvokeParams>();
+            const auto& tensors  = data_ctx.tensors;
+            auto argument_ptr    = conv_ptr.MakeArgumentPointer(
```

Comment:
[Performance] AFAICS the most of computations of kernel arguments can be moved outside of both lambdas (Invoker and InvokerFactory). These arguments may be computed right in GetSolution (or in InvokerFactory) and captured by value by InvokerFactory (or by Invoker). This works because the Solver produces distinct instance of InvokerFactory in each Solution, and then the InvokerCache produces and maintains distinct instance of Invoker per each ProblemConfig. When the library wants to run a Solution, it runs Invoker, which simply initializes pointers to the actual I/O buffers and calls the kernel -- much less work than now.

Please look how `opArgs` is used in `MakeImplGemmDynamicBackwardDataXdlopsNHWCInvokerFactory()`, for example. In #680 you can find some review comments related to this topic.

Reply:
I agree, however, the challenge is that since the `composable_kernel` API requires `unique_ptr` which makes capture by value difficult. For more details [please see here](https://github.com/ROCmSoftwarePlatform/composable_kernel/blob/91d8b7d67ae9dbf8a6e691ea3e17c0b9705c6ba7/library/include/ck/library/host/host_interface.hpp)

Reply:
Please let me know if you have any suggestions to improve it.